src/observation.jl

```diff
@@ -0,0 +1,199 @@
+"""
+    default_obsdim(data)
+
+The specify the default obsdim for a specific type of data.
+Defaults to `ObsDim.Undefined()`
+"""
+default_obsdim(data) = nothing
+default_obsdim(A::AbstractArray) = ndims(A)
+default_obsdim(tup::Tuple) = map(default_obsdim, tup)
+
+"""
+     datasubset(data, [idx], [obsdim])
+
+Return a lazy subset of the observations in `data` that correspond
+to the given `idx`. No data should be copied except of the
+indices. Note that `idx` can be of type `Int` or `AbstractVector`.
+Both options must be supported by a custom type.
+
+If it makes sense for the type of `data`, `obsdim` can be used
+to disptach on which dimension of `data` denotes the observations.
+See `?ObsDim`.
+"""
+function datasubset end
+
+#    getobs(data, [idx], [obsdim])
+#
+# Return the observations corresponding to the observation-index
+# `idx`. Note that `idx` can be of type `Int` or `AbstractVector`.
+# Both options must be supported by a custom type.
+#
+# The returned observation(s) should be in the form intended to
+# be passed as-is to some learning algorithm. There is no strict
+# interface requirement on how this "actual data" must look like.
+# Every author behind some custom data container can make this
+# decision him-/herself. We do, however, expect it to be consistent
+# for `idx` being an integer, as well as `idx` being an abstract
+# vector, respectively.
+#
+# If it makes sense for the type of `data`, `obsdim` can be used
+# to disptach on which dimension of `data` denotes the observations.
+# See `?ObsDim`
+#
+# This function is implemented in MLDataPattern
+function getobs end
+getobs(data, idx; obsdim = default_obsdim(data)) = getindex(data, idx)
```

Comment:
Hmmm, sorry for the inaccuracy. We don't need to define this line here, this one is meaningless.

Instead, what we need to do is to refresh MLDataPattern into keyword API. More specifically, remove the dispatch-based `getobs` methods in https://github.com/JuliaML/MLDataPattern.jl/blob/ac4227ebb8ddefb87d480c6bd80f784c7971aac6/src/container.jl#L1-L144. If dispatches are needed, it should be implemented as internal `_getobs` methods.

Reply:
Yeah, we can always use `_getobs` internally whenever `data` needs dispatch on the type of `obsdim`. My comment was more that this needs to be done for every `data` container that requires such behavior. But this is probably a rare use case beyond `Tuple`.

So should I remove the `getindex` default?

Reply:
I feel it's less safe a fallback to directly drop the `obsdim` here; this fallback does not handle either Tuple or AbstractArray case correctly.

So yeah I think it's better to remove the `getindex` default here.